enhancements/network/ovn-multi-external-gateway.md

```diff
@@ -0,0 +1,183 @@
+* `k8s.ovn.org/routing-external-gws`
+* `k8s.ovn.org/external-gw-pod-ips`
+
+Pod annotations:
+* `k8s.ovn.org/routing-namespaces`
+* `k8s.ovn.org/routing-network`
+* `k8s.ovn.org/bfd-enabled`
+
+The implementation will also include a new controller that will watch cluster wide for instances of this CRD and apply the changes to the OVN network configuration as it is currently done by the annotation based logic.
+
+The controller will use the OpenShift release version as well as the existence of the annotations in the namespace and pods to apply the changes to the network configuration:
+
+| **OpenShift version** | **Annotations exist** | **CR exists** | **Action**                                        |
+|-----------------------|-----------------------|---------------|---------------------------------------------------|
+| 4.13                  | No                    | Yes           | Use CR information to configure external gateways |
+| 4.13                  | Yes                   | Yes           | Use annotations to configure external gateways    |
+| 4.13                  | Yes                   | No            | Use annotations to configure external gateway     |
+| 4.13                  | No                    | No            | Nothing to do                                     |
+| 4.14                  | No                    | Yes           | Use CR information to configure external gateways |
+| 4.14                  | Yes                   | Yes           | Use CR information to configure external gateways |
+| 4.14                  | Yes                   | No            | Use annotations to configure external gateways    |
+| 4.14                  | No                    | No            | Nothing to do                                     |
+
+The current annotation based logic will be enhanced as part of this change to support the logic described in the previous table.
+
+### Workflow Description
+The workflow follows the behavior of a namespaced CR.
+
+
+### API Extensions
+
+A new namespaced-scoped CRD is created to represent the multiple external gateways configuration. The annotations are mapped as following:
+* `k8s.ovn.org/routing-external-gws` is a namespace specific annotation and is represented in the CRD as `routingExternalGateways`. The values represent a slice of strings containing the IPs used for external traffic. 
+This annotation works in conjunction with `k8s.ovn.org/external-gw-pod-ips` and is used to capture the IPs that are stored in conntrack.
+* `k8s.ovn.org/routing-namespaces` is mapped as `routingNamespaces` and it also contains a slice of the namespaces linked to this gateway configuration.
+* `k8s.ovn.org/routing-network` is mapped as `routingNetwork` and defines the multus network configuration that matches the same name in the pod that will be used for external traffic. If the field is empty or undefined, the pod's IP will be used instead. This matches the current behavior used with annotations.
```

Comment:
So today each exgw pod can configure this field to indicate to OVNK which IP to use from their multus network attachment. That means if we have 2 exgws, Pod A and Pod B, they could potentially have different routing-network values and still serve the same client namespace/pods. If we only allow for a singular value here, it means any pods matching the selector (lets assume Pod A and Pod B) would have to use the same network. Should we make this a plural slice of potential networks, and OVNK will just search each pod for one of the values? Or should we make it more fine grained with multiple selectors like:

```yaml
spec:
  podGateways:
    - network: blue
      podSelector:
        - matchLabels:
          - external-gateway: true
     - network: red
       podSelector:
         - matchLabels:
           - external-gateway2: true
```

Similarly I think we allow bfd to be enabled on a per pod basis (but not on a per static IP defined in the namespace basis). So maybe we need to have granularity there as well and separate it from statically assigned IPs. Like:

```yaml
kind: MultipleExternalGateway
apiVersion: k8s.ovn.org/v1
metadata:
  name: multiple-external-gateways-configuration
  namespace: default
spec:
  staticGateways:  ["9.0.0.1", "9.0.0.2"]
  staticGatewaysBFD: true
  podGateways:
    - network: blue
      bfdEnabled: true
      podSelector:
        - matchLabels:
          - external-gateway: true
     - network: red
       podSelector:
         - matchLabels:
           - external-gateway2: true
status:
  externalGatewayPodIPs: ["10.10.1.2","10.10.1.5"]
```

Reply:
Makes sense about BFD being per pod and being able to have multiple selectors.

Reply:
What is the purpose of `staticGatewaysBFD` in the manifest? If it's a deployment/pod specific feature, does it make sense to apply to the whole manifest?

Reply:
so that is there to signal to use BFD or not with the gateways specified in the staticGateways. Today we only allow you to configure BFD on the namespace for the entirety of all static IPs defined in the namespace annotation (not on an individual basis). But for pods acting as external gateways, we allow you to annotate those individually. In hindsight this wasn't a very consistent design decision :) So, since we do not want to change the functional behavior I think we need to still allow for individual BFD selection for pod external gateways, but not for statically assigned IPs (previously done in the namespace annotation).

Does that make sense?

@fedepaol can you double check me here please.

Reply:
yeah, checking the code it does what you mentioned. On a second thought, it might make more sense to allow defining the BFD per external IP rather than as a whole as it is currently done 

```yaml
staticExternalGateways:
  - ip: "192.168.1.1"
    bfdEnabled: true
  - ip: "10.1.1.1"
```
I don't mind keeping it for backwards compatibility if it's worth keeping it, but I'm happy to rework/drop stuff that doesn't make sense or is not used/deprecated.

Reply:
yeah it would be nice to allow that, i was just trying to avoid any change in functionality so that there would not be extra work to do there. I don't think it is  a huge effort though to change this behavior.

Reply:
I'm proposing to define 2 CRDs, one for each type of traffic, but keeping the similar high level fields (`from`,`nextHops`). Here is an update of the example posted earlier in this thread:

```yaml
apiVersion: k8s.ovn.org/v1beta1
kind: AdminPolicyBasedExternalRoute
metadata:
  name: honeypotting
spec:
  policies:
  ## gateway example
  - from:
      namespaceSelector:
        matchLabels:
            multiple_gws: true
    nexthops:
      - ip: "172.18.0.2"
        bfdEnabled: true
      - ip: "10.244.0.3"
        bfdEnabled: false
      - podSelector:
          matchLabels:
            external-gateway: true
        bfdEnabled: true
        namespaceSelector:
          matchLabels:
            gateway: true
        networkAttachmentName: sriov1
```